### PR TITLE
Idempotency

### LINKED ISSUE #130: CSI Node issue on unmount volume
/kind bug

**What happened?**
We experience one time, a pod stuck on terminating state and the volume was not unmount properly.

**Idea**
Implement a stress test that mount and unmount multiple times to see if we can reproduce it

### LINKED ISSUE #162: Fix Non-Idempotent functions
/kind bug

**What happened?**
The [CSI specification](https://github.com/container-storage-interface/spec/blob/master/spec.md) impose that all functions must be idempotent.

Issue #130 shows that `NodePublishVolume` was not idempotent.

** What to do ?**
We need to check the idenpotancy of all functions.

### PR DESCRIPTION
**Pull request by outscale-mdr**
Closes #130 
Closes #162

### CODE REVIEW

**Comment by outscale-mdr** [comment]
Also closes #162

File: pkg/driver/node.go [hunk 1]
```diff
@@ -544,12 +552,19 @@ func (d *nodeService) nodePublishVolumeForBlock(req *csi.NodePublishVolumeReques
...
-	klog.V(5).Infof("NodePublishVolume [block]: mounting %s at %s", source, target)
-	if err := d.mounter.Mount(source, target, "", mountOptions); err != nil {
-		if removeErr := os.Remove(target); removeErr != nil {
-			return status.Errorf(codes.Internal, "Could not remove mount target %q: %v", target, removeErr)
+	isMounted, err := d.isMounted(target)
+	if err != nil {
+		return status.Errorf(codes.Internal, "Could not check if %q is mounted: %v", target, err)
+	}
+
+	if !isMounted {
```
Comment (jerome-jutteau) [inline pkg/driver/node.go:560]: maybe
```
if isMounted {
    return nil
}
...
```

File: pkg/driver/node.go [hunk 2]
```diff
@@ -576,12 +591,19 @@ func (d *nodeService) nodePublishVolumeForFileSystem(req *csi.NodePublishVolumeR
...
-	klog.V(5).Infof("NodePublishVolume: mounting %s at %s with option %s as fstype %s", source, target, mountOptions, fsType)
-	if err := d.mounter.Mount(source, target, fsType, mountOptions); err != nil {
-		if removeErr := os.Remove(target); removeErr != nil {
-			return status.Errorf(codes.Internal, "Could not remove mount target %q: %v", target, err)
+	isMounted, err := d.isMounted(target)
+	if err != nil {
+		return status.Errorf(codes.Internal, "Could not check if %q is mounted: %v", target, err)
+	}
+
+	if !isMounted {
```
Comment (jerome-jutteau) [inline pkg/driver/node.go:599]: same :)

File: pkg/driver/node.go [hunk 3]
```diff
@@ -660,3 +682,45 @@ func hasMountOption(options []string, opt string) bool {
...
 }
+
+// isMounted checks if target is mounted. It does NOT return an error if target
+// doesn't exist.
+func (d *nodeService) isMounted(target string) (bool, error) {
+	/*
+		Checking if it's a mount point using IsLikelyNotMountPoint. There are three different return values,
+		1. true, err when the directory does not exist or corrupted.
+		2. false, nil when the path is already mounted with a device.
+		3. true, nil when the path is not mounted with any device.
```
Comment (jerome-jutteau) [inline pkg/driver/node.go:693]: Maybe this description can be placed above function declaration.
Also: 
> 1. true, err when the directory does not exist or corrupted

Seems to provide an error but this seems contrary to
> It does NOT return an error if target doesn't exist.

**Review by jerome-jutteau — APPROVED** [review]

File: pkg/driver/node.go (hunk 1 above)
Comment (outscale-mdr) [reply pkg/driver/node.go]: Yes, better for readibility

File: pkg/driver/node.go (hunk 3 above)
Comment (outscale-mdr) [reply pkg/driver/node.go:697]: The comments explain the output of `IsLikelyNotMountPoint`, not the output of `isMounted`.

**Pull request merged**